pallets/gear/src/lib.rs

```diff
@@ -1379,72 +1487,27 @@ pub mod pallet {
         ) -> DispatchResultWithPostInfo {
             let who = ensure_signed(origin)?;
 
-            // Check that provided `gas_limit` value does not exceed the block gas limit
-            ensure!(
-                gas_limit <= BlockGasLimitOf::<T>::get(),
-                Error::<T>::GasLimitTooHigh
-            );
-
-            let numeric_value: u128 = value.unique_saturated_into();
-            let minimum: u128 = CurrencyOf::<T>::minimum_balance().unique_saturated_into();
-
-            // Check that provided `value` equals 0 or greater than existential deposit
-            ensure!(
-                0 == numeric_value || numeric_value >= minimum,
-                Error::<T>::ValueLessThanMinimal
-            );
-
-            let schedule = T::Schedule::get();
-
-            ensure!(
-                code.len() as u32 <= schedule.limits.code_len,
-                Error::<T>::CodeTooLarge
-            );
-
-            let code = Code::try_new(code, schedule.instruction_weights.version, |module| {
-                schedule.rules(module)
-            })
-            .map_err(|e| {
-                log::debug!("Code failed to load: {:?}", e);
-                Error::<T>::FailedToConstructProgram
-            })?;
-
-            ensure!(
-                code.code().len() as u32 <= schedule.limits.code_len,
-                Error::<T>::CodeTooLarge
-            );
+            Self::check_gas_limit_and_value(gas_limit, value)?;
 
-            let code_and_id = CodeAndId::new(code);
+            let code_and_id = Self::check_code(code)?;
+            if T::CodeStorage::exists(code_and_id.code_id()) {
```

Comment:
I think it's wrong to return an error if someone tries to upload the same program. He will anyway pay for that (because of message size). Moreover, it won't lead to writing again the code to the code storage by the logic of `set_code_with_metadata`.

Reply:
Well, I don't agree with the fact that we must revert transaction with existing code as a parameter. In this case, we are like making users follow some determined flow of program creation. In this case `create_program` turns out to be not a feature, but a "must call, if code exists". I guess better remove the existence check.

Reply:
Initially my comment was about don't set code if it already exists, but do not fail extrinsic.

I see them like:
- Upload code:
   - If code does not exist, we set it and extrinsic OK
   - If code already exist, we don't set it again and extrinsic OK (we will not also charge money for code existence further)
- Upload program:
   - If code exists, we use it and extrinsic OK
   - If code doesn't exists, we set it and extrinsic OK
   - If program exists, then extrinsic ERR

Weight mechanism of frame support also allows us to charge for only happened case (if code wasn't set in upload_program and/or upload code, then funds for setting will be returned, but its for future)

Reply:
I see no reason to restrict some extrinsic with it.

Atm imagine the situation,
Two users found some code and decided to upload program with it. They made sure that there is no such code in storage and they use the different salts. Sending it really close in time will make one of the extrinsic fail, that wasn't expected by user. That's strange

Reply:
> `Bob` doesn't know why `PROGRAM` with `SALT_BOB` has not been created.

We should continue execution if code exists and should create program with it, charging less money from `Bob` due to absence of code saving

Reply:
> > Weight mechanism of frame support also allows us to charge for only happened case (if code wasn't set in upload_program and/or upload code, then funds for setting will be returned, but its for future)
> 
> I believe we can just do this for the weight corrections inside extrinsic
> 
> ```rust
> fn upload_program(..) -> DispatchWithPostInfo {
>     if code_exists_but_created_new_program {
>         return Ok(Some(
>             <T as Config>::WeightInfo::upload_program(0, salt.len() as u32),
>         ).into());
>     }
> }
> ```

Yeah, I meant exactly this way. We just will need two benches of different storage states for the extrinsic